File: SwiftDemoAppUITests/Pages/LoginPage.swift
import XCTest

class LoginPage {
    
    let app: XCUIApplication
    
    init(app: XCUIApplication) {
        self.app = app
    }
    
    var usernameField: XCUIElement { app.textFields["username-field"]}
    var passwordField: XCUIElement { app.secureTextFields["password-field"]}
    var loginButton: XCUIElement { app.buttons["login-button"]}
    var errorLabel: XCUIElement { app.staticTexts["error-label"]}
    
    @discardableResult func enterUsername(_ username: String) -> LoginPage{
        usernameField.tap()
        usernameField.typeText(username)
        return self
    }
    
    @discardableResult func enterPassword(_ password: String) -> LoginPage{
        passwordField.tap()
        passwordField.typeText(password)
        return self
    }
    
    @discardableResult func tapLogin() -> LoginPage{
        loginButton.tap()
        return self
    }
    
    func login(username: String, password: String) {
        enterUsername(username)
        enterPassword(password)
        tapLogin()
    }
    
    func assertLoginScreenVisible() {
        XCTAssertTrue(loginButton.waitForExistence(timeout: 5), "Login screen should be visible")
    }
    
    func assertErrormessage(_ expected: String) {
        sleep(2)
        XCTAssertEqual(
            errorLabel.label,
            expected,
            "Error message mismatch"
        )
    }
}

File: pages/login_page.py
from pages.base_page import BasePage
from appium.webdriver.webdriver import WebDriver
from appium.webdriver.common.appiumby import AppiumBy

class LoginPage(BasePage):

    def __init__(self, driver: WebDriver):
        super().__init__(driver)

    USERNAME_FIELD = (AppiumBy.ACCESSIBILITY_ID, "username-field")
    PASSWORD_FIELD = (AppiumBy.ACCESSIBILITY_ID, "password-field")
    LOGIN_BUTTON = (AppiumBy.ACCESSIBILITY_ID, "login-button")
    ERROR_LABEL = (AppiumBy.ACCESSIBILITY_ID, "error-label")

    def enter_username(self, username: str):
        self.type_text(self.USERNAME_FIELD, username)

    def enter_password(self, password: str):
        self.type_text(self.PASSWORD_FIELD, password)

    def click_login(self):
        self.tap(self.LOGIN_BUTTON)

    def login(self,username: str, password: str):
        self.enter_username(username)
        self.enter_password(password)
        self.click_login()
